Add hiv resistance detection (#559)
nf-core/viralrecon: docs/output.md

    - [ABACAS](#abacas)
    - [PlasmidID](#plasmidid)
    - [Assembly QUAST](#assembly-quast)
  - [Illumina: HIV resistance detection](#illumina-hiv-resistance-detection)
    - [sierra-local](#sierra-local)

>     - [liftoff](#liftoff)

Reviewer: ```suggestion
    - [Liftoff](#Liftoff)
```

>     - [codfreq](#codfreq)

Reviewer: ```suggestion
    - [CodFreq](#CodFreq)
```

![MultiQC - QUAST contig counts](images/mqc_quast_plot.png)

## Illumina: HIV resistance detection

This output is only generated if `--perform_hiv_resistance` param is activated.

### sierra-local

[**sierra-local**](https://github.com/PoonLab/sierra-local/) is a Python3 implementation of the [Stanford University HIV Drug Resistance Database](https://hivdb.stanford.edu/) (HIVdb) [Sierra web service](https://hivdb.stanford.edu/page/webservice/).

> It allows to generate **HIV-1 drug resistance predictions locally**, without the need to transmit patient data over a network. This provides full control over **data provenance**, **security**, and **regulatory compliance**.

Reviewer: Not sure if you need to emphasize some of the text?

Author: True, I can keep it simple!

> _Sierra-local computes drug resistance predictions directly from consensus HIV-1 sequences, producing JSON-formatted output that includes key resistance-associated mutations and drug susceptibility scores._

> > Within the HIV module of `nf-core/viralrecon`, sierra-local is executed using the consensus .FASTA file generated for each sample.

Reviewer: ```suggestion
> Within the HIV module of `nf-core/viralrecon`, sierra-local is executed using the consensus FASTA file generated for each sample.
```

<details markdown="1">
<summary>Output files</summary>

- `variants/<CALLER>/consensus/<CONSENSUS_CALLER>/sierra-local`

>   - `*_resistance.json`: Sierralocal resistance output JSON file. This JSON file summarizes the HIV drug resistance interpretation from a consensus sequence of the PR, RT and IN genes of HIV. It includes sequence metadata (`inputSequence`) (FASTA header and SHA-512 checksum), quality control messages (`validationResults`), detected mutations (`alignedGeneSequences`), and inferred drug resistance levels (`drugResistance`). This output contains information similar to that provided by the HIVdb's HTML output, and a complete document of this HTML output can be found in HIVdb's website [here](https://hivdb.stanford.edu/page/xml-2009-spec/).

Reviewer: ```suggestion
  - `*_resistance.json`: sierra-local resistance output JSON file. This JSON file summarizes the HIV drug resistance interpretation from a consensus sequence of the PR, RT and IN genes of HIV. It includes sequence metadata (`inputSequence`) (FASTA header and SHA-512 checksum), quality control messages (`validationResults`), detected mutations (`alignedGeneSequences`), and inferred drug resistance levels (`drugResistance`). This output contains information similar to that provided by the HIVdb's HTML output, and a complete document of this HTML output can be found in HIVdb's website [here](https://hivdb.stanford.edu/page/xml-2009-spec/).
```

**NB:** The value of `<CALLER>` and `<CONSENSUS_CALLER>` in the output directory name above is determined by the `--variant_caller` and `--consensus_caller` parameters (Default: 'ivar' for both params when '--profile test_hiv').

</details>

> ### liftoff

Reviewer: ```suggestion
### Liftoff
```

> ### liftoff

Reviewer: I think you have already explained why the default genome is `codfreq` in the test dataset, right? Do you think it’s necessary to document it here as well?

Author: I don't think we need to explain default parameters in output documentation.

Only performed when the `--genome` is not `codfreq`. **[Liftoff](https://github.com/agshumate/Liftoff)** is a tool that accurately maps annotations in GFF or GTF between assemblies of the same, or closely-related species. Unlike current coordinate lift-over tools which require a pre-generated “chain” file as input, Liftoff is a standalone tool that takes two genome assemblies and a reference annotation as input and outputs an annotation of the target genome. Using the provided reference genome and the `.fasta` and `.gff`, annotation files are processed with to transfer the HIV gene annotations to the user’s specific reference genome.

Once the new `.gff` file is produced by Liftoff, an additional annotation step is performed. The variant caller outputs (based on **nucleotide-level variants**) are re-annotated using this updated `.gff`, generating a **secondary annotation layer**.

<details markdown="1">
<summary>Output files</summary>

- `variants/<CALLER>/consensus/<CONSENSUS_CALLER>/`
  - `liftoff/`

>     - `<reference_basename>.gff`: Annotation GFF with the HIV genes specific annotations needed for the analysis.

Reviewer: What do you mean HIV genes? It's unclear to me.

Author: I think Features is more unclear, which is what the F from GFF means... It's a gene/feature coordinate file for HIV only in this pipeline. I can re-write the sentence to be les redundant, but I think Annotated genes is the most clear way to describe it.

    - `<reference_basename>.unmapped.txt`: Liftoff file containing unmapped features.
  - `liftoff_variants_long_table.csv`: Long format table similar to `variants_long_table.csv` for the `<reference_basename>.gff` annotation file needed for HIV resistance detection analysis.

**NB:** The value of `<CALLER>` and `<CONSENSUS_CALLER>` in the output directory name above is determined by the `--variant_caller` and `--consensus_caller` parameters (Default: 'ivar' for both callers when using '--profile test_hiv').

</details>

> ### codfreq

Reviewer: ```suggestion
### CodFreq
```

| `count`               | Total reads supporting that codon                                   |
| `total_quality_score` | Qualituy score of that codon assigned by codfreq                    |
| `aa_codon`            | If the codon is multiple of 3, the corresponding aminoacid sequence |

### Custom HIV Resistance Reports

Using both **sierra-local**'s `*_resistance.json` and **custom codfreq**'s `*.codfreq` files, an [in-house script](../bin/resistance_tables.py) is used to generate HIV resistance reports needed for clinical purposes.

- `variants/<CALLER>/consensus/<CONSENSUS_CALLER>/resistance_reports`
  - `*_mutation_table.csv`: CSV table with the mutations found by sierra-local with codon coverage and mutation frequencies integrated. It contains the following sexteen columns:
    - `Sample_name`: Name of the sample given by the samplesshet file.
    - `Gene_name`: Name of the gene containing the mutation. Only detects PR, RT and IN.
    - `Mutations`: Detected aminoacid mutation in this format `I72V` as `<reference_aminoacid><codon_position><alternative_aminoacid>`.
    - `Mutations_type`: Mutation's primary type. Can be `Major Mutation`, `Accesory Mutation` or `Other`.

>     - `Mutations_comments`: Resistance comments from sierra-local software for that mutation. Can be empty if no comments where detected. Only major mutations contain comments.

Reviewer: Did we decide to put this column here or as the last column in the table?

Author: They wanted it close to the mutation.

    - `isInsertion`: If the mutation is an insertion `True`. Else `False`.
    - `isDeletion`: If the mutation is a deletion `True`. Else `False`.
    - `isApobecMutation`: If the mutation is an Apobec mutation found in `apobecs.csv` file `True`. Else `False`.
    - `isApobecDRM`: If the mutation is an Apobec DRM mutation found in `apobec_drms.json` file `True`. Else `False`.
    - `isUnusual`: If the mutation is an unusual mutation found in `rx-all_subtype-all.csv` file `True`. Else `False`.
    - `isSDRM`:If the mutation is a SDRM mutation found in `sdrms_hiv1.csv` file `True`. Else `False`.
    - `hasStop`: If the mutation has a Stop codon `True`. Else `False`.
    - `Mutation_AF`: Aminoacid mutation allele frequency. Calculated: (alternative_aminoacid total coverage) / (codon total coverage)
    - `Coverage`: Codon total coverage. Used to calculate `Mutation_AF`.

>     - `INDEL>5%`: If INDELs represent >=5% of the codon reads/coverage `True`. Else `False`. Can be used to detect posible artifacts in a position if `isInsertion` and `isDeletion` are `False`.

Reviewer: What do you mean "`isInsertion` and `isDeletion` are `False`."here? I thought we check indels proportion of each mutation.

Author: If the detected mutation is an INDEL `isInsertion` and `isDeletion` will be True, and then it will represent more than the 5% of the coverage, so `INDEL>5%` won't represent any artifact, but the real case of the mutation. If the mutation is not an INDEL and there's not any INDEL mutation detected in that codon position (thus `isInsertion` and `isDeletion` are `False`) BUT this `INDEL>5%` is True, then it can help to detect possible artifact.

      - `INSTI`: Integrase Strand Transfer Inhibitors (IN gene asociated)
    - `Drugs`: Name of the drug for which the suscteptibility is being evaluated. Can be:
      - PR gene related (Protease Inhibitors):
        - `ATV/r`: atazanavir/r
        - `DRV/r`: darunavir/r
        - `LPV/r`: lopinavir/r
        - `FPV/r`: fosamprenavir/r
        - `IDV/r`: indinavir/r
        - `NFV`: nelfinavir
        - `SQV/r`: saquinavir/r
        - `TPV/r`: tipranavir/r
      - RT gene related:
        - Nucleoside Reverse Transcriptase Inhibitors (NRTI)
          - `ABC`: abacavir
          - `AZT`: zidovudine
          - `FTC`: emtricitabine
          - `3TC`: lamivudine
          - `TDF`: tenofovir
          - `D4T`: stavudine
          - `DDI`: didanosine
        - Non-nucleoside Reverse Transcriptase Inhibitors (NNRTI)
          - `DOR`: doravirine
          - `EFV`: efavirenz
          - `ETR`: etravirine
          - `NVP`: nevirapine
          - `RPV`: rilpivirine
          - `DPV`: dapivirine
      - IN gene related (Integrase Strand Transfer Inhibitors):
        - `BIC`: bictegravir
        - `CAB`: cabotegravir
        - `DTG`: dolutegravir
        - `EVG`: elvitegravir

>         - `RAL`: raltegravir

Reviewer: Nice! You inlcuded the map of drugs  abbreviations and full name here.
